### PR TITLE
holochain_net: TestSuite

### PR DESCRIPTION
**Pull request by ddd-mtl**
## TestSuite foundation for `holochain_net`
Created a P2pNode which is a struct holding a P2pNetwork and the receiver channel.
A P2pNode has a constructor for every 'type' of p2p module:
- rust_mock: MockSystem from 'net' crate
- ipc_mock:  n3h mock mode
- normal: n3h hack mode

The TestSuite triggers the test functions for every type of p2p module.
a test function takes only two P2pNodes as input, so doesn't care about the type of config.
Currently the test functions are listed in a vec:
- setup_normal
- send_test
- dht_test
- meta_test

TestSuite coded from test_bin_ipc which is now main.rs as the previous main.rs was not doing anything.
Kept test_bin_mock_net for now but is not really useful anymore.

MockSystem now sends PeerConnected messages on receiving TrackApp message.
MockWorker's endpoint is the network's name

(some changes are just code that I moved around)

### Misc
- Moved MockSystem to its own file.
- Added lots of comments for clarifyng net code.

### Minor renames
P2pConfig::named_mock_config => P2pConfig::named_mock_as_string
NetConnection trait => NetSend trait
IpcNetWorker.state => last_known_state

### CODE REVIEW

**Comment by zippy** [comment]
still some CI probs here...

**Comment by ddd-mtl** [comment]
@zippy Since I couldn't merged I continued work on this branch so now your current review is out of date. Waiting for a new review :)

File: net/src/ipc_net_worker.rs [hunk 1]
```diff
@@ -232,13 +182,87 @@ impl IpcNetWorker {
...
+        self.ipc_relay.send(data)?;
+        Ok(())
+    }
+
+    /// do some upkeep on the internal worker
+    /// IPC server state handling / magic
+    fn tick(&mut self) -> NetResult<bool> {
+        let mut has_done_something = false;
+
+        // WTF?
```
Comment (neonphog) [inline net/src/ipc_net_worker.rs]: WTF? : ) lol

could say something like

```rust
// if we have yet to issue a state request to the networking process, issue one now
```

Comment (ddd-mtl) [reply net/src/ipc_net_worker.rs]: oops, forgot to remove that one :p

**Pull request merged**